share/man/man4/aps.4
.\"	$NetBSD$
.\"	$OpenBSD: aps.4,v 1.7 2007/05/31 19:19:49 jmc Exp $
.\"
.\" Copyright (c) 2005 Jonathan Gray <jsg@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd July 13, 2011
.Dt APS 4
.Os
.Sh NAME
.Nm aps
.Nd ThinkPad Active Protection System accelerometer
.Sh SYNOPSIS
.Cd "aps0 at isa? port 0x1600"
.Sh DESCRIPTION
The
.Nm
driver provides support for several sensors found in some ThinkPad laptops.
.Pp
The sensors currently exposed via the
.Xr envsys 4
interface are:
.Bl -column "Sensor        " "Units    " "Typical" -offset indent
.It Sy "Sensor        " Ta Sy "Units    " Ta Sy "Typical Use"
.It Li "x-acceleration" Ta "integer" Ta "X-axis acceleration"
.It Li "y-acceleration" Ta "integer" Ta "Y-axis acceleration"
.It Li "x-variable" Ta "integer" Ta "Weighted X acceleration?"
.It Li "y-variable" Ta "integer" Ta "Weighted Y acceleration?"
.It Li "temperature 1" Ta "degC" Ta "Unknown temperature"
.It Li "temperature 2" Ta "degC" Ta "Unknown temperature"
.It Li "keyboard active" Ta "boolean" Ta "Keyboard activity"
.It Li "mouse active" Ta "boolean" Ta "Mouse activity"
.It Li "lid open" Ta "boolean" Ta "Lid state"
.El
.Sh SEE ALSO
.Xr envsys 4 ,
.Xr hpacel 4 ,
.Xr thinkpad 4 ,
.Xr envstat 8
.Sh HISTORY
The
.Nm
driver first appeared in
.Ox 3.8
and was then ported to
.Nx
5.0.
.Sh AUTHORS
The
.Nm
driver was written by
.An Jonathan Gray Aq Mt jsg@openbsd.org .
.Sh CAVEATS
Few issues can be mentioned.
.Bl -bullet
.It
The
.Nm
driver does not maintain state and subsequently does not take
evasive action when it thinks the hard drive is in danger.
Possible actions would include spinning down the hard drive
in case excessive tremor is detected by the sensors.
.It
The Y axis on X40 and possibly other models seems to be inverted.
It is unknown how to distinguish between different versions of the
accelerometer to compensate for this in the driver at this time.
.It
The sensor values are refreshed every 0.5 seconds.
Because no protection measures are taken,
this is unnecessary and may have a negative effect on battery life.
.It
As IBM provides no documentation, it is not known what all the available
sensors are used for.
.El


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

sys/dev/isa/aps.c
/*	$NetBSD$	*/
/*	$OpenBSD: aps.c,v 1.15 2007/05/19 19:14:11 tedu Exp $	*/
/*	$OpenBSD: aps.c,v 1.17 2008/06/27 06:08:43 canacar Exp $	*/
/*
 * Copyright (c) 2005 Jonathan Gray <jsg@openbsd.org>
 * Copyright (c) 2008 Can Erkin Acar <canacar@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * A driver for the ThinkPad Active Protection System based on notes from
 * http://www.almaden.ibm.com/cs/people/marksmith/tpaps.html
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/device.h>
#include <sys/kernel.h>
#include <sys/callout.h>
#include <sys/module.h>

#include <sys/bus.h>

#include <dev/sysmon/sysmonvar.h>

#include <dev/isa/isareg.h>
#include <dev/isa/isavar.h>

#if defined(APSDEBUG)
#define DPRINTF(x)		do { printf x; } while (0)
#else
#define DPRINTF(x)
#endif


/*
 * EC interface on Thinkpad Laptops, from Linux HDAPS driver notes.
 * From Renesans H8S/2140B Group Hardware Manual
 * http://documentation.renesas.com/eng/products/mpumcu/rej09b0300_2140bhm.pdf
 *
 * EC uses LPC Channel 3 registers TWR0..15
 */

/* STR3 status register */
#define APS_STR3		0x04

#define APS_STR3_IBF3B	0x80	/* Input buffer full (host->slave) */
#define APS_STR3_OBF3B	0x40	/* Output buffer full (slave->host)*/
#define APS_STR3_MWMF	0x20	/* Master write mode */
#define APS_STR3_SWMF	0x10	/* Slave write mode */


/* Base address of TWR registers */
#define APS_TWR_BASE		0x10
#define APS_TWR_RET		0x1f

/* TWR registers */
#define APS_CMD			0x00
#define APS_ARG1		0x01
#define APS_ARG2		0x02
#define APS_ARG3		0x03
#define APS_RET			0x0f

/* Sensor values */
#define APS_STATE		0x01
#define	APS_XACCEL		0x02
#define APS_YACCEL		0x04
#define APS_TEMP		0x06
#define	APS_XVAR		0x07
#define APS_YVAR		0x09
#define APS_TEMP2		0x0b
#define APS_UNKNOWN		0x0c
#define APS_INPUT		0x0d

/* write masks for I/O, send command + 0-3 arguments*/
#define APS_WRITE_0		0x0001
#define APS_WRITE_1		0x0003
#define APS_WRITE_2		0x0007
#define APS_WRITE_3		0x000f

/* read masks for I/O, read 0-3 values (skip command byte) */
#define APS_READ_0		0x0000
#define APS_READ_1		0x0002
#define APS_READ_2		0x0006
#define APS_READ_3		0x000e

#define APS_READ_RET		0x8000
#define APS_READ_ALL		0xffff

/* Bit definitions for APS_INPUT value */
#define APS_INPUT_KB		(1 << 5)
#define APS_INPUT_MS		(1 << 6)
#define APS_INPUT_LIDOPEN	(1 << 7)

#define APS_ADDR_SIZE		0x1f

struct sensor_rec {
	uint8_t 	state;
	uint16_t	x_accel;
	uint16_t	y_accel;
	uint8_t 	temp1;
	uint16_t	x_var;
	uint16_t	y_var;
	uint8_t 	temp2;
	uint8_t 	unk;
	uint8_t 	input;
};

enum aps_sensors {
        APS_SENSOR_XACCEL = 0,
        APS_SENSOR_YACCEL,
        APS_SENSOR_XVAR,
        APS_SENSOR_YVAR,
        APS_SENSOR_TEMP1,
        APS_SENSOR_TEMP2,
        APS_SENSOR_KBACT,
        APS_SENSOR_MSACT,
        APS_SENSOR_LIDOPEN,
        APS_NUM_SENSORS
};

struct aps_softc {
	bus_space_tag_t sc_iot;
	bus_space_handle_t sc_ioh;
	bool sc_bus_space_valid;

	struct sysmon_envsys *sc_sme;
	envsys_data_t sc_sensor[APS_NUM_SENSORS];
	struct callout sc_callout;

	struct sensor_rec aps_data;
};

static int 	aps_match(device_t, cfdata_t, void *);
static void 	aps_attach(device_t, device_t, void *);
static int	aps_detach(device_t, int);

static int 	aps_init(struct aps_softc *);
static int	aps_read_data(struct aps_softc *);
static void 	aps_refresh_sensor_data(struct aps_softc *);
static void 	aps_refresh(void *);
static int	aps_do_io(bus_space_tag_t, bus_space_handle_t,
			  unsigned char *, int, int);
static bool 	aps_suspend(device_t, const pmf_qual_t *);
static bool 	aps_resume(device_t, const pmf_qual_t *);

CFATTACH_DECL_NEW(aps, sizeof(struct aps_softc),
	      aps_match, aps_attach, aps_detach, NULL);

/* properly communicate with the controller, writing a set of memory
 * locations and reading back another set  */
static int
aps_do_io(bus_space_tag_t iot, bus_space_handle_t ioh,
	  unsigned char *buf, int wmask, int rmask)
{
	int bp, stat, n;

	DPRINTF(("aps_do_io: CMD: 0x%02x, wmask: 0x%04x, rmask: 0x%04x\n",
	    buf[0], wmask, rmask));

	/* write init byte using arbitration */
	for (n = 0; n < 100; n++) {
		stat = bus_space_read_1(iot, ioh, APS_STR3);
		if (stat & (APS_STR3_OBF3B | APS_STR3_SWMF)) {
			bus_space_read_1(iot, ioh, APS_TWR_RET);
			continue;
		}
		bus_space_write_1(iot, ioh, APS_TWR_BASE, buf[0]);
		stat = bus_space_read_1(iot, ioh, APS_STR3);
		if (stat & (APS_STR3_MWMF))
			break;
		delay(1);
	}

	if (n == 100) {
		DPRINTF(("aps_do_io: Failed to get bus\n"));
		return 1;
	}

	/* write data bytes, init already sent */
	/* make sure last bye is always written as this will trigger slave */
	wmask |= APS_READ_RET;
	buf[APS_RET] = 0x01;

	for (n = 1, bp = 2; n < 16; bp <<= 1, n++) {
		if (wmask & bp) {
			bus_space_write_1(iot, ioh, APS_TWR_BASE + n, buf[n]);
			DPRINTF(("aps_do_io:  write %2d 0x%02x\n", n, buf[n]));
		}
	}

	for (n = 0; n < 100; n++) {
		stat = bus_space_read_1(iot, ioh, APS_STR3);
		if (stat & (APS_STR3_OBF3B))
			break;
		delay(5 * 100);
	}

	if (n == 100) {
		DPRINTF(("aps_do_io: timeout waiting response\n"));
		return 1;
	}
	/* wait for data available */
	/* make sure to read the final byte to clear status */
	rmask |= APS_READ_RET;

	/* read cmd and data bytes */
	for (n = 0, bp = 1; n < 16; bp <<= 1, n++) {
		if (rmask & bp) {
			buf[n] = bus_space_read_1(iot, ioh, APS_TWR_BASE + n);
			DPRINTF(("aps_do_io:  read %2d 0x%02x\n", n, buf[n]));
		}
	}

	return 0;
}

static int
aps_match(device_t parent, cfdata_t match, void *aux)
{
	struct isa_attach_args *ia = aux;
	bus_space_tag_t iot = ia->ia_iot;
	bus_space_handle_t ioh;
	unsigned char iobuf[16];
	int iobase;
	uint8_t cr;

	/* Must supply an address */
	if (ia->ia_nio < 1)
		return 0;

	if (ISA_DIRECT_CONFIG(ia))
		return 0;

	if (ia->ia_io[0].ir_addr == ISA_UNKNOWN_PORT)
		return 0;

	iobase = ia->ia_io[0].ir_addr;

	if (bus_space_map(iot, iobase, APS_ADDR_SIZE, 0, &ioh)) {
		aprint_error("aps: can't map i/o space\n");
		return 0;
	}


	/* See if this machine has APS */

	/* get APS mode */
	iobuf[APS_CMD] = 0x13;
	if (aps_do_io(iot, ioh, iobuf, APS_WRITE_0, APS_READ_1)) {
		bus_space_unmap(iot, ioh, APS_ADDR_SIZE);
		return 0;
	}

	/*
	 * Observed values from Linux driver:
	 * 0x01: T42
	 * 0x02: chip already initialised
	 * 0x03: T41
	 * 0x05: T61
	 */

	cr = iobuf[APS_ARG1];

	bus_space_unmap(iot, ioh, APS_ADDR_SIZE);
	DPRINTF(("aps: state register 0x%x\n", cr));

	if (iobuf[APS_RET] != 0 || cr < 1 || cr > 5) {
		DPRINTF(("aps0: unsupported state %d\n", cr));
		return 0;
	}

	ia->ia_nio = 1;
	ia->ia_io[0].ir_size = APS_ADDR_SIZE;
	ia->ia_niomem = 0;
	ia->ia_nirq = 0;
	ia->ia_ndrq = 0;

	return 1;
}

static void
aps_attach(device_t parent, device_t self, void *aux)
{
	struct aps_softc *sc = device_private(self);
	struct isa_attach_args *ia = aux;
	int iobase, i;

	sc->sc_iot = ia->ia_iot;
	iobase = ia->ia_io[0].ir_addr;

	callout_init(&sc->sc_callout, 0);
	callout_setfunc(&sc->sc_callout, aps_refresh, sc);

	if (bus_space_map(sc->sc_iot, iobase, APS_ADDR_SIZE, 0, &sc->sc_ioh)) {
		aprint_error(": can't map i/o space\n");
		return;
	}
	sc->sc_bus_space_valid = true;

	aprint_naive("\n");
	aprint_normal(": Thinkpad Active Protection System\n");

	if (aps_init(sc)) {
		aprint_error_dev(self, "failed to initialize\n");
		goto out;
	}

	/* Initialize sensors */
#define INITDATA(idx, unit, string)					\
	sc->sc_sensor[idx].units = unit;				\
	strlcpy(sc->sc_sensor[idx].desc, string,			\
	    sizeof(sc->sc_sensor[idx].desc));

	INITDATA(APS_SENSOR_XACCEL, ENVSYS_INTEGER, "x-acceleration");
	INITDATA(APS_SENSOR_YACCEL, ENVSYS_INTEGER, "y-acceleration");
	INITDATA(APS_SENSOR_TEMP1, ENVSYS_STEMP, "temperature 1");
	INITDATA(APS_SENSOR_TEMP2, ENVSYS_STEMP, "temperature 2");
	INITDATA(APS_SENSOR_XVAR, ENVSYS_INTEGER, "x-variable");
	INITDATA(APS_SENSOR_YVAR, ENVSYS_INTEGER, "y-variable");
	INITDATA(APS_SENSOR_KBACT, ENVSYS_INDICATOR, "keyboard active");
	INITDATA(APS_SENSOR_MSACT, ENVSYS_INDICATOR, "mouse active");
	INITDATA(APS_SENSOR_LIDOPEN, ENVSYS_INDICATOR, "lid open");

	sc->sc_sme = sysmon_envsys_create();

	for (i = 0; i < APS_NUM_SENSORS; i++) {

		sc->sc_sensor[i].state = ENVSYS_SVALID;

		if (sc->sc_sensor[i].units == ENVSYS_INTEGER)
			sc->sc_sensor[i].flags = ENVSYS_FHAS_ENTROPY;

		if (sysmon_envsys_sensor_attach(sc->sc_sme,
			&sc->sc_sensor[i])) {
			sysmon_envsys_destroy(sc->sc_sme);
			goto out;
		}
	}
        /*
         * Register with the sysmon_envsys(9) framework.
         */
	sc->sc_sme->sme_name = device_xname(self);
	sc->sc_sme->sme_flags = SME_DISABLE_REFRESH;

	if ((i = sysmon_envsys_register(sc->sc_sme))) {
		aprint_error_dev(self,
		    "unable to register with sysmon (%d)\n", i);
		sysmon_envsys_destroy(sc->sc_sme);
		goto out;
	}

	if (!pmf_device_register(self, aps_suspend, aps_resume))
		aprint_error_dev(self, "couldn't establish power handler\n");

	/* Refresh sensor data every 0.5 seconds */
	callout_schedule(&sc->sc_callout, (hz) / 2);

	return;

out:
	bus_space_unmap(sc->sc_iot, sc->sc_ioh, APS_ADDR_SIZE);
}

static int
aps_init(struct aps_softc *sc)
{
	unsigned char iobuf[16];

	/* command 0x17/0x81: check EC */
	iobuf[APS_CMD] = 0x17;
	iobuf[APS_ARG1] = 0x81;

	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_1, APS_READ_3))
		return 1;
	if (iobuf[APS_RET] != 0 ||iobuf[APS_ARG3] != 0)
		return 1;

	/* Test values from the Linux driver */
	if ((iobuf[APS_ARG1] != 0 || iobuf[APS_ARG2] != 0x60) &&
	    (iobuf[APS_ARG1] != 1 || iobuf[APS_ARG2] != 0))
		return 1;

	/* command 0x14: set power */
	iobuf[APS_CMD] = 0x14;
	iobuf[APS_ARG1] = 0x01;

	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_1, APS_READ_0))
		return 1;

	if (iobuf[APS_RET] != 0)
		return 1;

	/* command 0x10: set config (sample rate and order) */
	iobuf[APS_CMD] = 0x10;
	iobuf[APS_ARG1] = 0xc8;
	iobuf[APS_ARG2] = 0x00;
	iobuf[APS_ARG3] = 0x02;

	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_3, APS_READ_0))
		return 1;

	/* command 0x11: refresh data */
	iobuf[APS_CMD] = 0x11;
	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_0, APS_READ_1))
		return 1;
	if (iobuf[APS_ARG1] != 0)
		return 1;

	return 0;
}

static int
aps_detach(device_t self, int flags)
{
	struct aps_softc *sc = device_private(self);

        callout_stop(&sc->sc_callout);
        callout_destroy(&sc->sc_callout);

	if (sc->sc_sme)
		sysmon_envsys_unregister(sc->sc_sme);
	if (sc->sc_bus_space_valid == true)
		bus_space_unmap(sc->sc_iot, sc->sc_ioh, APS_ADDR_SIZE);

	return 0;
}

static int
aps_read_data(struct aps_softc *sc)
{
	unsigned char iobuf[16];

	iobuf[APS_CMD] = 0x11;
	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_0, APS_READ_ALL))
		return 1;

	sc->aps_data.state = iobuf[APS_STATE];
	sc->aps_data.x_accel = iobuf[APS_XACCEL] + 256 * iobuf[APS_XACCEL + 1];
	sc->aps_data.y_accel = iobuf[APS_YACCEL] + 256 * iobuf[APS_YACCEL + 1];
	sc->aps_data.temp1 = iobuf[APS_TEMP];
	sc->aps_data.x_var = iobuf[APS_XVAR] + 256 * iobuf[APS_XVAR + 1];
	sc->aps_data.y_var = iobuf[APS_YVAR] + 256 * iobuf[APS_YVAR + 1];
	sc->aps_data.temp2 = iobuf[APS_TEMP2];
	sc->aps_data.input = iobuf[APS_INPUT];

	return 0;
}

static void
aps_refresh_sensor_data(struct aps_softc *sc)
{
	int64_t temp;

	if (aps_read_data(sc)) {
		printf("aps0: read data failed\n");
		return;
	}

	sc->sc_sensor[APS_SENSOR_XACCEL].value_cur = sc->aps_data.x_accel;
	sc->sc_sensor[APS_SENSOR_YACCEL].value_cur = sc->aps_data.y_accel;

	if (sc->aps_data.temp1 == 0xff)
		sc->sc_sensor[APS_SENSOR_TEMP1].state = ENVSYS_SINVALID;
	else {
		/* convert to micro (mu) degrees */
		temp = sc->aps_data.temp1 * 1000000;	
		/* convert to kelvin */
		temp += 273150000; 
		sc->sc_sensor[APS_SENSOR_TEMP1].value_cur = temp;
		sc->sc_sensor[APS_SENSOR_TEMP1].state = ENVSYS_SVALID;
	}

	if (sc->aps_data.temp2 == 0xff)
		sc->sc_sensor[APS_SENSOR_TEMP2].state = ENVSYS_SINVALID;
	else {
		/* convert to micro (mu) degrees */
		temp = sc->aps_data.temp2 * 1000000;	
		/* convert to kelvin */
		temp += 273150000; 
		sc->sc_sensor[APS_SENSOR_TEMP2].value_cur = temp;
		sc->sc_sensor[APS_SENSOR_TEMP2].state = ENVSYS_SVALID;
	}

	sc->sc_sensor[APS_SENSOR_XVAR].value_cur = sc->aps_data.x_var;
	sc->sc_sensor[APS_SENSOR_YVAR].value_cur = sc->aps_data.y_var;
	sc->sc_sensor[APS_SENSOR_KBACT].value_cur =
	    (sc->aps_data.input &  APS_INPUT_KB) ? 1 : 0;
	sc->sc_sensor[APS_SENSOR_MSACT].value_cur =
	    (sc->aps_data.input & APS_INPUT_MS) ? 1 : 0;
	sc->sc_sensor[APS_SENSOR_LIDOPEN].value_cur =
	    (sc->aps_data.input & APS_INPUT_LIDOPEN) ? 1 : 0;
}

static void
aps_refresh(void *arg)
{
	struct aps_softc *sc = arg;

	aps_refresh_sensor_data(sc);
	callout_schedule(&sc->sc_callout, (hz) / 2);
}

static bool
aps_suspend(device_t dv, const pmf_qual_t *qual)
{
	struct aps_softc *sc = device_private(dv);

	callout_stop(&sc->sc_callout);

	return true;
}

static bool
aps_resume(device_t dv, const pmf_qual_t *qual)
{
	struct aps_softc *sc = device_private(dv);
	unsigned char iobuf[16];

	/*
	 * Redo the init sequence on resume, because APS is 
	 * as forgetful as it is deaf.
	 */

	/* get APS mode */
	iobuf[APS_CMD] = 0x13;
	if (aps_do_io(sc->sc_iot, sc->sc_ioh, iobuf, APS_WRITE_0, APS_READ_1)
	    || aps_init(sc))
		aprint_error_dev(dv, "failed to wake up\n");
	else
		callout_schedule(&sc->sc_callout, (hz) / 2);

	return true;
}

MODULE(MODULE_CLASS_DRIVER, aps, NULL);

#ifdef _MODULE
#include "ioconf.c"
#endif

static int
aps_modcmd(modcmd_t cmd, void *opaque)
{
	switch (cmd) {
	case MODULE_CMD_INIT:
#ifdef _MODULE
		return config_init_component(cfdriver_ioconf_aps,
		    cfattach_ioconf_aps, cfdata_ioconf_aps);
#else
		return 0;
#endif
	case MODULE_CMD_FINI:
#ifdef _MODULE
		return config_fini_component(cfdriver_ioconf_aps,
		    cfattach_ioconf_aps, cfdata_ioconf_aps);
#else
		return 0;
#endif
	default:
		return ENOTTY;
	}
}
